### PR TITLE
Get expr planners when creating new planner

### LINKED ISSUE #11477: ExprPlanner not propagated to SqlToRel
### Describe the bug

When upgrading to DataFusion 40, I kept on encountering this error:

```
Internal error: Expected a simplified result, but none was found.
```

This happened when parsing a array literal. It seems in order to parse it, we need to register the proper `ExprPlanner` on the `SessionState`. We use `SessionState::default()`, which registers it for us. But then it's not being propagated to `SqlToRel`, so it fails. (Note the `planners: vec![]`)

https://github.com/apache/datafusion/blob/f204869ff55bb3e39cf23fc0a34ebd5021e6773f/datafusion/sql/src/planner.rs#L206-L213

### To Reproduce

_No response_

### Expected behavior

I'm not sure I understand the design, but it seems odd we don't propagate these. At the very least, we should provide a warning in the docs.

### Additional context

Current workaround:

```
for planner in context_provider.state.expr_planners() {
    sql_to_rel = sql_to_rel.with_user_defined_planner(planner.clone());
}
```

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #11477 .

### CODE REVIEW

**Comment by jayzhan211** [comment]
We may even remove planners in SqlToRel 🤔 ?

```rust
/// SQL query planner
pub struct SqlToRel<'a, S: ContextProvider> {
    pub(crate) context_provider: &'a S,
    pub(crate) options: ParserOptions,
    pub(crate) normalizer: IdentNormalizer,
    /// user defined planner extensions
    pub(crate) planners: Vec<Arc<dyn ExprPlanner>>,
}
```

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -196,12 +193,6 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         Self::new_with_options(context_provider, ParserOptions::default())
     }
 
-    /// add an user defined planner
-    pub fn with_user_defined_planner(mut self, planner: Arc<dyn ExprPlanner>) -> Self {
```
Comment (jayzhan211) [inline datafusion/sql/src/planner.rs:200]: I think we can register planner in session state

File: datafusion/expr/src/planner.rs [hunk 2]
```diff
@@ -60,6 +60,9 @@ pub trait ContextProvider {
         not_impl_err!("Recursive CTE is not implemented")
     }
 
+    /// Getter for expr planners
+    fn get_expr_planners(&self) -> Vec<Arc<dyn ExprPlanner>>;
```
Comment (alamb) [inline datafusion/expr/src/planner.rs:64]: I spent some time trying to figure out if we could avoid this Vec cloning all the time -- with

```suggestion
    fn get_expr_planners(&self) -> &[Arc<dyn ExprPlanner>] {&[]}
```

I think this is possible -- I made https://github.com/jayzhan211/arrow-datafusion/pull/3 to show how it looks

**Review by alamb — APPROVED** [review]
Thanks @jayzhan211  this makes sense to me. 

I have a suggestion to avoid creating Vec's for all accesses to extension planners. Let me know what you think: https://github.com/jayzhan211/arrow-datafusion/pull/3

I also wonder if we should add a test case to avoid this kind of issue from happening again (whatever API Lance and Delta.rs are using clearly isn't covered by our existing tests)

cc @wjones127 and @rtyler  (who also mentioned this https://github.com/apache/datafusion/pull/11180#discussion_r1667729599)

**Review by wjones127 — COMMENTED** [review]
Thanks. This seems like a good simplification. I also agree that we could make `get_expr_planners` return a slice.

File: datafusion/core/src/execution/session_state.rs [hunk 3]
```diff
@@ -1597,12 +1583,20 @@ impl SessionStateDefaults {
...
+/// This is used so the SQL planner can access the state of the session without
+/// having a direct dependency on the [`SessionState`] struct (and core crate)
 struct SessionContextProvider<'a> {
     state: &'a SessionState,
     tables: HashMap<String, Arc<dyn TableSource>>,
 }
 
 impl<'a> ContextProvider for SessionContextProvider<'a> {
+    fn get_expr_planners(&self) -> &[Arc<dyn ExprPlanner>] {
+        &self.state.expr_planners
```
Comment (jayzhan211) [inline datafusion/core/src/execution/session_state.rs:1597]: I remove `expr_planners` field in `SessionContextProvider` and get it from state directly

Comment (alamb) [reply datafusion/core/src/execution/session_state.rs:1597]: Nice! I tried to figure out how to do this previously and got confused about a lock or something. This is very nice 👌

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @jayzhan211 and @wjones127

**Pull request merged**